### PR TITLE
Fix valgrind issues

### PR DESCRIPTION
**Pull request by galsalomon66**

### CODE REVIEW

File: include/s3select_functions.h [hunk 1]
```diff
@@ -151,18 +170,27 @@ class __function : public base_statement
...
 };
 
 
+  void s3select_functions::clean()
+  {
+    for(auto d : __all_query_functions)
+    {
+      if (d->is_function())
+      {
+        dynamic_cast<__function*>(d)->impl()->__call_destructor();
```
Comment (mattbenjamin) [inline include/s3select_functions.h:188]: adding a new syntax "__call_destructor" that just calls "->~func()" is superfluous;  just call the dtor--every C++ programmer is familiar with it.

File: include/s3select_functions.h [hunk 2]
```diff
@@ -151,18 +170,27 @@ class __function : public base_statement
...
+      }
+      d->__call_destructor();
```
Comment (mattbenjamin) [inline include/s3select_functions.h:190]: while I'm looking at this code section, I see a bunch of member variables whose names begin with double-underscore;  I don't think that is helpful.  Just give member variables ordinary names.

File: include/s3select_oper.h [hunk 3]
```diff
@@ -17,6 +17,93 @@ namespace bsc = BOOST_SPIRIT_CLASSIC_NS;
...
+  //==================================
+  inline T* _Allocate(size_t _Count, T*)
+  {
+    // allocate storage for _Count elements of type T
+
+    pointer res = (pointer)(__pool + __pool_loc);
+
+    __pool_loc += sizeof(T) * _Count;
+    //alignment
+    size_t _algn = ((size_t)(__pool_loc) % sizeof(char*));
```
Comment (mattbenjamin) [inline include/s3select_oper.h:55]: ok, c++....

File: include/s3select_oper.h [hunk 4]
```diff
@@ -17,6 +17,93 @@ namespace bsc = BOOST_SPIRIT_CLASSIC_NS;
...
+{
+public:
+  typedef size_t size_type;
+  typedef T* pointer;
+  typedef const T* const_pointer;
+  size_t __pool_loc;
+  char* __pool;
+
+  //only ONE pool,not allocated dynamically; main assumption, caller knows in advance its memory limitations.
+  char __buff__[pool_sz];
```
Comment (mattbenjamin) [inline include/s3select_oper.h:38]: ok, sane design, but lose all the double-underscores ;)

File: include/s3select_functions.h (hunk 1 above)
Comment (galsalomon66) [reply include/s3select_functions.h:188]: ok, will rename.

the s3-select-function framework is calling to the base virtual function(__cal_dest...) , which call base::~dtor() (and that triggered the derived dtor())
meaning, in that way it's transparent to new derived object ( such as s3-select operator / function )

**Pull request merged**